File: src/main/java/implementslegendkt/mod/vaultjp/screen/view/BackgroundViewDSL.java
package implementslegendkt.mod.vaultjp.screen.view;

import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.vertex.PoseStack;
import implementslegendkt.mod.vaultjp.screen.DecentScreen;
import implementslegendkt.mod.vaultjp.screen.View;
import implementslegendkt.mod.vaultjp.screen.ViewInteractor;
import net.minecraft.client.renderer.GameRenderer;
import net.minecraft.client.renderer.Rect2i;
import net.minecraft.resources.ResourceLocation;
import oshi.util.tuples.Pair;

import java.util.ArrayList;
import java.util.function.Supplier;

public class BackgroundViewDSL implements View {

    public Supplier<ResourceLocation> texture;
    public Supplier<Rect2i> srcRect;
    public Supplier<Pair<Integer,Integer>> atlasSize;
    public Supplier<Pair<Integer,Integer>> pos;

    public static class BackgroundInteractor implements ViewInteractor<BackgroundViewDSL> {

        private ArrayList<BackgroundViewDSL> views = new ArrayList<>();

        @Override
        public void clear() {
            views.clear();
        }

        @Override
        public void addView(BackgroundViewDSL dsl) {
            views.add(dsl);
        }

        public <S extends DecentScreen<S,?>> void renderViews(S screen, PoseStack stack, int cursorX, int cursorY) {
            screen.fillGradient(stack, 0, 0, screen.width, screen.height, -1072689136, -804253680);
            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.ScreenEvent.BackgroundDrawnEvent(screen, stack));
            for (var view:views) {
                RenderSystem.setShader(GameRenderer::getPositionTexShader);
                RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
                RenderSystem.setShaderTexture(0, view.texture.get());
                var rect = view.srcRect.get();
                var pos = view.pos.get();
                var atlas = view.atlasSize.get();
                screen.blit(stack, pos.getA(), pos.getB(), screen.getBlitOffset(), rect.getX(), rect.getY(),rect.getWidth(), rect.getHeight(),atlas.getA(),atlas.getB());
            }
        }

        @Override
        public <S extends DecentScreen<S,?>> void click(S screen, int cursorX, int cursorY, int key) {}

    }
}

File: src/main/java/implementslegendkt/mod/screenlegends/view/BackgroundViewDSL.kt
package implementslegendkt.mod.screenlegends.view

import com.mojang.blaze3d.systems.RenderSystem
import com.mojang.blaze3d.vertex.PoseStack
import implementslegendkt.mod.screenlegends.DecentScreen
import implementslegendkt.mod.screenlegends.View
import implementslegendkt.mod.screenlegends.ViewInteractor
import implementslegendkt.mod.screenlegends.forEnabledViews
import net.minecraft.client.gui.GuiComponent
import net.minecraft.client.renderer.GameRenderer
import net.minecraft.client.renderer.Rect2i
import net.minecraft.resources.ResourceLocation
import net.minecraftforge.client.event.ScreenEvent.BackgroundDrawnEvent
import net.minecraftforge.common.MinecraftForge

class BackgroundViewDSL : View {
    var texture: ()->ResourceLocation? = { null }
    var srcRect: ()->Rect2i? = { null }
    var atlasSize: ()->Pair<Int, Int>? = { null }
    var pos: ()->Pair<Int, Int> = { 0 to 0}

    override var enabled: () -> Boolean = {true}

    class BackgroundInteractor : ViewInteractor<BackgroundViewDSL> {
        override val views = ArrayList<BackgroundViewDSL>()

        override fun clear() {
            views.clear()
        }

        override fun addView(dsl: BackgroundViewDSL) {
            views.add(dsl)
        }

        override fun <S : DecentScreen<*, *>> renderViews(screen: S, stack: PoseStack?, cursorX: Int, cursorY: Int) {
            screen.fillGradient(stack!!, 0, 0, screen.width, screen.height, -1072689136, -804253680)
            MinecraftForge.EVENT_BUS.post(BackgroundDrawnEvent(screen, stack))
            forEnabledViews view@{view->
                RenderSystem.setShader { GameRenderer.getPositionTexShader() }
                RenderSystem.setShaderColor(1.0f, 1.0f, 1.0f, 1.0f)
                view.texture()?.let { RenderSystem.setShaderTexture(0, it) }
                val rect = view.srcRect()?:return@view
                val pos = view.pos()
                val atlas = view.atlasSize()?:return@view
                GuiComponent.blit(
                    stack,
                    pos.first,
                    pos.second,
                    screen.getBlitOffset(),
                    rect.x.toFloat(),
                    rect.y.toFloat(),
                    rect.width,
                    rect.height,
                    atlas.first,
                    atlas.second
                )
            }
        }

        override fun <S : DecentScreen<*, *>> click(screen: S, cursorX: Int, cursorY: Int, key: Int) {}
    }

}
